src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs

```diff
@@ -874,6 +874,55 @@ private static void ClearAndReturn(ArraySegment<byte> rented)
             }
         }
 
+        internal void WritePropertyName(int index, Utf8JsonWriter writer)
+        {
+            CheckNotDisposed();
+
+            WritePropertyName(_parsedData.Get(index - DbRow.Size), writer);
+        }
+
+        private void WritePropertyNameNew(in DbRow row, Utf8JsonWriter writer)
+        {
+            // To be determined.
+            // This method is ~10% faster than the original WritePropertyName
+            // when the property name contains escaped/Unicode characters.
+
+            Debug.Assert(row.TokenType == JsonTokenType.PropertyName);
+            int loc = row.Location;
+            int length = row.SizeOrLength;
+            ReadOnlySpan<byte> rawName = _utf8Json.Slice(loc, length).Span;
+
+            int firstBackSlashIndex = rawName.IndexOf(JsonConstants.BackSlash);
+
+            if (firstBackSlashIndex < 0)
+            {
+                writer.WritePropertyName(rawName);
+                return;
+            }
+
+            // If the name needs unescaping
+
+            ArraySegment<byte> rented = default;
+
+            Span<byte> utf8Unescaped = length <= JsonConstants.StackallocByteThreshold ?
+                stackalloc byte[JsonConstants.StackallocByteThreshold] :
+                (rented = new ArraySegment<byte>(ArrayPool<byte>.Shared.Rent(length), 0, length));
```

Comment:
Why the ArraySegment allocation?

Reply:
> Why the ArraySegment?

The original `JsonDocument.WritePropertyName` & `JsonDocument.UnescapeString` uses `ArraySegment<byte>` to mark rented arrays.. 
Meanwhile `TextEquals` and the like uses `byte[]?`.